### PR TITLE
WIP: Check that hooks registered by a module have a related listener implemented

### LINKED ISSUE #19230: Notify when a module registers to a hook but implements no hook listener
Following #15702

#### Is your feature request related to a problem?

When a module registers to a hook using `$this->registerHook()`, no check is performed to make sure that said module implements a hook method to listen to those hooks. This could happen for example when there's a typing mistake in the hook name.

#### Describe the solution you'd like

In debug mode, throw an exception like this when that happens:

> Hook with the name %hookName% has been registered by %moduleClassName%, but the corresponding method %functionName% has not been defined in the Module class.

### PR DESCRIPTION
**Pull request by matthieu-rolland**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | When installing a module, we want to check that all registered hooks have corresponding listeners implemented in the module, otherwise throw an exception;
| Type?         | improvement
| Category?     | CO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Fixes #19230
| How to test?  | Install a module that registers an existing hook without implementing a listener for it. In debug mode.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/21450)

### CODE REVIEW

File: classes/Hook.php [hunk 1]
```diff
@@ -582,6 +582,18 @@ public static function registerHook($module_instance, $hook_name, $shop_list = n
                 return false;
             }
 
+            // check that hook listener is implemented by the module
+            $expectedHookListenerMethod = sprintf('hook%s', ucfirst($hook_name));
```
Comment (matks) [inline classes/Hook.php]: I think - I'm not sure - that hook callbacks are case-insensitive so we might need to check both `sprintf('hook%s', ucfirst($hook_name))` and `sprintf('hook%s', lcfirst($hook_name))` 🤔  ?

Comment (kpodemski) [reply classes/Hook.php]: They are ineed case-insensitive, you can have `hookdisplayprodctadditionalinfo` 👿 and it still does work

@matthieu-rolland maybe you could use this method:
https://github.com/PrestaShop/PrestaShop/blob/develop/classes/Hook.php#L386 ?

Comment (matthieu-rolland) [reply classes/Hook.php]: Indeed, looks like the methods `getAllKnownNames` and `getHookAliasesFor` take care of the different possible hook names.

Fixed 👍

**Comment by kpodemski** [comment]
@PrestaShop/prestashop-maintainers 

`1) LegacyTests\Integration\classes\module\ModuleGetPossibleHooksListTest::testGetRightListForModule
PrestaShopException: Hook with the name displayHome has been registered by bankwire, but the corresponding method hookDisplayHome has not been defined in the Module class.`

https://github.com/PrestaShop/PrestaShop/blob/develop/tests-legacy/Integration/classes/module/ModuleGetPossibleHooksListTest.php#L45

Shall we skip/delete this test? Changing legacy test to `ps_wirepayment` is an option? Not sure how we approach legacy tests

**Review by atomiix — DISMISSED** [review]

**Review by PierreRambaud — CHANGES_REQUESTED** [review]
Make sure you can install PrestaShop :wink:

**Comment by matthieu-rolland** [comment]
> Make sure you can install PrestaShop 😉

yep, this PR can't be merged as long as the modules I fixed are not released...

I put the PR in WIP

**Comment by matks** [comment]
@matthieu-rolland maybe we can put a list of the modules waiting for a fix in the github issue, so we can monitor the progress? 😉

**Comment by PierreRambaud** [comment]
Hi @matthieu-rolland ,

I'm not sure the work is still in progress. Feel free to reopen it if you think you're able to work again on it.

Thanks!

**Pull request closed without merging**